File: apps/atlas/test/go_to_state_test.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

import 'package:atlas_core/atlas_core.dart';
import 'package:atlas_geo/atlas_geo.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:atlas/go_to/go_to_state.dart';

AtlasCoordinate _coord(double latitude, double longitude) {
  return AtlasCoordinate(latitude: latitude, longitude: longitude);
}

void main() {
  test('fresh state has no active target', () {
    final state = GoToState();
    expect(state.isActive, isFalse);
    expect(state.target, isNull);
    expect(state.distanceKmTo(_coord(0.0, 0.0)), isNull);
    expect(state.bearingDegTo(_coord(0.0, 0.0)), isNull);
  });

  test('activate preserves id coordinate and label', () {
    final state = GoToState();
    state.activate(
      id: 'wp-000001',
      latitude: 45.0,
      longitude: -93.0,
      label: 'Alpha',
    );
    expect(state.isActive, isTrue);
    expect(state.target?.id, 'wp-000001');
    expect(state.target?.latitude, 45.0);
    expect(state.target?.longitude, -93.0);
    expect(state.target?.label, 'Alpha');
  });

  test('activate rejects invalid coordinates instead of inventing', () {
    final state = GoToState();
    expect(
      () => state.activate(
        id: 'wp-000001',
        latitude: 91.0,
        longitude: 0.0,
        label: 'Bad',
      ),
      throwsA(isA<AtlasRejectionException>()),
    );
    expect(state.isActive, isFalse);
  });

  test('distance follows the golden track-length vector', () {
    final state = GoToState();
    state.activate(id: 'w2', latitude: 0.0, longitude: 1.0, label: 'East');
    expect(state.distanceKmTo(_coord(0.0, 0.0)), closeTo(111.19508, 0.001));
  });

  test('bearing runs current position to target', () {
    final state = GoToState();
    state.activate(id: 'w2', latitude: 0.0, longitude: 1.0, label: 'East');
    expect(state.bearingDegTo(_coord(0.0, 0.0)), closeTo(90.0, 1e-9));
    expect(state.bearingDegTo(_coord(0.0, 1.0)), isNull);
  });

  test('coincident fix yields zero distance and undefined bearing', () {
    final state = GoToState();
    state.activate(id: 'w1', latitude: 45.0, longitude: -93.0, label: 'Here');
    expect(state.distanceKmTo(_coord(45.0, -93.0)), 0.0);
    expect(state.bearingDegTo(_coord(45.0, -93.0)), isNull);
  });

  test('missing fix yields unavailable navigation values', () {
    final state = GoToState();
    state.activate(id: 'w1', latitude: 45.0, longitude: -93.0, label: 'Far');
    expect(state.distanceKmTo(null), isNull);
    expect(state.bearingDegTo(null), isNull);
  });

  test('clear removes target and tolerates repeat calls', () {
    final state = GoToState();
    state.activate(id: 'w1', latitude: 45.0, longitude: -93.0, label: 'Far');
    state.clear();
    expect(state.isActive, isFalse);
    expect(state.target, isNull);
    state.clear();
    expect(state.isActive, isFalse);
  });

  test('reactivation replaces the previous target', () {
    final state = GoToState();
    state.activate(id: 'w1', latitude: 45.0, longitude: -93.0, label: 'One');
    state.activate(id: 'w2', latitude: 46.0, longitude: -94.0, label: 'Two');
    expect(state.target?.id, 'w2');
    expect(state.target?.label, 'Two');
  });
}

File: apps/atlas-android/app/src/test/java/com/sovereignatlas/atlas/goto/GoToStateTest.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

package com.sovereignatlas.atlas.goto

import com.sovereignatlas.atlas.geo.AtlasCoordinate
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

final class GoToStateTest {
    private fun active(): GoToState {
        val state = GoToState()
        state.activate(
            id = "wp-000001",
            latitude = 45.0,
            longitude = -93.0,
            label = "Alpha",
        )
        return state
    }

    @Test
    fun inactiveByDefault() {
        val state = GoToState()
        assertFalse(state.isActive())
        assertNull(state.targetOrNull())
        assertNull(state.distanceKmTo(AtlasCoordinate(latitude = 0.0, longitude = 0.0)))
        assertNull(state.bearingDegTo(AtlasCoordinate(latitude = 0.0, longitude = 0.0)))
    }

    @Test
    fun activateSnapshotsTarget() {
        val state = active()
        assertTrue(state.isActive())
        assertEquals("wp-000001", state.targetOrNull()?.id)
        assertEquals("Alpha", state.targetOrNull()?.label)
        assertEquals(45.0, state.targetOrNull()?.latitude)
    }

    @Test
    fun activateRejectsInvalidCoordinate() {
        val state = GoToState()
        try {
            state.activate(id = "x", latitude = 91.0, longitude = 0.0, label = "bad")
            throw AssertionError("latitude 91 must be rejected")
        } catch (error: Exception) {
            assertFalse(state.isActive())
        }
    }

    @Test
    fun clearDeactivatesIdempotently() {
        val state = active()
        var count = 0
        state.addListener { count++ }
        state.clear()
        assertFalse(state.isActive())
        state.clear()
        assertEquals(1, count)
    }

    @Test
    fun distanceAndBearingAreLive() {
        val state = active()
        val fix = AtlasCoordinate(latitude = 44.0, longitude = -93.0)
        val distance = state.distanceKmTo(fix)
        assertTrue(distance != null && distance > 100.0 && distance < 120.0)
        assertEquals(0.0, state.bearingDegTo(fix)!!, 0.5)
    }

    @Test
    fun nullFixYieldsNull() {
        val state = active()
        assertNull(state.distanceKmTo(null))
        assertNull(state.bearingDegTo(null))
    }

    @Test
    fun coincidentBearingIsNull() {
        val state = active()
        val fix = AtlasCoordinate(latitude = 45.0, longitude = -93.0)
        assertEquals(0.0, state.distanceKmTo(fix)!!, 0.0)
        assertNull(state.bearingDegTo(fix))
    }

    @Test
    fun listenersNotified() {
        val state = GoToState()
        var count = 0
        state.addListener { count++ }
        state.activate(id = "a", latitude = 1.0, longitude = 1.0, label = "A")
        state.clear()
        assertEquals(2, count)
    }
}
